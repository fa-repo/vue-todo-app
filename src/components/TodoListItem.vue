<template>
  <li class="todo-item" :class="{ 'is-complete': isComplete }">
    <input
      class="completed"
      type="checkbox"
      @change="toggleTodo"
      v-model="isComplete"
    />
    <input class="note" type="text" @blur="updateTodo" v-model="note" />
    <button class="delete" @click="removeTodo">‚ùå</button>
  </li>
</template>

<script>
export default {
  name: "TodoListItem",
  props: {
    todo: Object,
    onDelete: Function,
    onUpdate: Function
  },
  data() {
    return {
      isComplete: this.todo.complete,
      note: this.todo.note
    };
  },
  methods: {
    removeTodo: function() {
      this.onDelete(this.todo.id);
    },
    toggleTodo: function() {
      this.onUpdate({ ...this.todo, complete: !this.todo.complete });
    },
    updateTodo: function() {
      this.onUpdate({ ...this.todo, note: this.note });
    }
  }
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
li {
  display: flex;
  width: 100%;
  text-align: left;
  border-bottom: thin solid #cccccc;
  position: relative;
  align-items: center;
}

.completed {
  font-size: 16px;
  margin: 0 10px;
}

.note {
  flex: 1;
  border: 0;
  margin: 0;
  padding: 10px;
  font-size: 14px;
  outline: 0;
  &:focus {
    z-index: 999;
    box-shadow: 0px 0px 0 2px #2196F3;
  }
}

.delete {
  right: 0;
  padding: 9px 8px 9px 12px;
  border: none;
  outline: 0;
  &:focus {
    z-index: 999;
    box-shadow: 0px 0px 0 2px #2196F3;
  }
}
</style>
