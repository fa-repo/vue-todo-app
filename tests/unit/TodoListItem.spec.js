import TodoListItem from "@/components/TodoListItem.vue";
import { mount } from "@vue/test-utils";

describe("TodoListItem", () => {
  it("Renders properly", () => {
    const propsData = {
      todo: { id: "id", note: "yawww", complete: true }
    };
    const wrapper = mount(TodoListItem, { propsData });

    expect(wrapper.isVueInstance()).toBe(true);
    expect(wrapper.element).toMatchSnapshot();
  });

  test("Calls onUpdate after editing note", () => {
    const propsData = {
      todo: { id: "id", note: "yawww", complete: true },
      onUpdate: jest.fn()
    };
    const wrapper = mount(TodoListItem, { propsData });

    wrapper.find(".note").setValue("some input");
    wrapper.find(".note").trigger("blur");

    expect(wrapper.vm.onUpdate).toHaveBeenCalledTimes(1);
  });

  test("Calls onUpdate after toggling note", () => {
    const propsData = {
      todo: { id: "id", note: "yawww", complete: true },
      onUpdate: jest.fn()
    };
    const wrapper = mount(TodoListItem, { propsData });

    wrapper.find(".completed").trigger("click");

    expect(wrapper.vm.onUpdate).toHaveBeenCalledTimes(1);
  });

  test("Calls onRemove after deleting note", () => {
    const propsData = {
      todo: { id: "id", note: "yawww", complete: true },
      onDelete: jest.fn()
    };
    const wrapper = mount(TodoListItem, { propsData });

    wrapper.find(".delete").trigger("click");

    expect(wrapper.vm.onDelete).toHaveBeenCalledTimes(1);
  });
});
