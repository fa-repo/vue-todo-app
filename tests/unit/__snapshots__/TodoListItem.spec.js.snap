// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListItem Calls onUpdate after editing note 1`] = `
<li
  class="todo-item is-complete"
>
  <input
    class="toggle"
    type="checkbox"
  />
   
  <input
    class="note"
    type="text"
  />
   
  <button
    class="delete"
  >
    ❌
  </button>
</li>
`;

exports[`TodoListItem Renders properly 1`] = `
<li
  class="todo-item is-complete"
>
  <input
    class="completed"
    type="checkbox"
  />
   
  <input
    class="note"
    type="text"
  />
   
  <button
    class="delete"
  >
    ❌
  </button>
</li>
`;
